### PR TITLE
A few cleanups for hir

### PR DESCRIPTION
**Pull request by ljedrz**
- prefer `if let` to `match` when only 1 branch matters
- `chain` iterable items that are looped over in sequence
- `sort_by_key` instead of `sort_by` when possible
- change cloning `map`s to `cloned()`
- use `unwrap_or_else` and `ok` when applicable
- a few other minor readability improvements
- whitespace fixes

### CODE REVIEW

File: src/librustc/hir/intravisit.rs [hunk 1]
```diff
@@ -123,9 +122,8 @@ impl<'this, 'tcx> NestedVisitorMap<'this, 'tcx> {
     /// e.g., item, impl-item.
     pub fn inter(self) -> Option<&'this Map<'tcx>> {
         match self {
-            NestedVisitorMap::None => None,
-            NestedVisitorMap::OnlyBodies(_) => None,
             NestedVisitorMap::All(map) => Some(map),
+            _ => None
```
Comment (michaelwoerister) [inline src/librustc/hir/intravisit.rs]: Usually we try to make match statements exhaustive in order to catch errors when new variants are added.

File: src/librustc/hir/print.rs [hunk 2]
```diff
@@ -848,9 +845,8 @@ impl<'a> State<'a> {
     }
 
     pub fn print_defaultness(&mut self, defaultness: hir::Defaultness) -> io::Result<()> {
-        match defaultness {
-            hir::Defaultness::Default { .. } => self.word_nbsp("default")?,
-            hir::Defaultness::Final => (),
+        if let hir::Defaultness::Default { .. } = defaultness {
+            self.word_nbsp("default")?
```
Comment (michaelwoerister) [inline src/librustc/hir/print.rs]: I'd keep this the old way too. It's good to see all the options listed.

File: src/librustc/hir/print.rs [hunk 3]
```diff
@@ -2341,14 +2331,12 @@ impl<'a> State<'a> {
...
-            hir::Constness::Const => self.word_nbsp("const")?,
+        if let hir::Constness::Const = header.constness {
+            self.word_nbsp("const")?
         }
 
-        match header.asyncness {
-            hir::IsAsync::NotAsync => {}
-            hir::IsAsync::Async => self.word_nbsp("async")?,
+        if let hir::IsAsync::Async = header.asyncness {
+            self.word_nbsp("async")?
```
Comment (michaelwoerister) [inline src/librustc/hir/print.rs]: The same here and `constness` above. It's a bit more verbose to list all variants, but also a lot more readable for someone who does not know the code base by heart.

**Comment by michaelwoerister** [comment]
Thanks for the PR, @ljedrz! I've lot some comments in the code.

**Comment by ljedrz** [comment]
@michaelwoerister thanks, comments addressed.

**Comment by michaelwoerister** [comment]
Thanks, @ljedrz! 

@bors r+

**Pull request merged**